### PR TITLE
Perf Investigation

### PR DESCRIPTION
**Pull request by subhoghoshX**
On first look, found two issues:
1. Creating a component inside the function body of another component creates a brand new component on every render, so React will update the *actual dom* thinking it's different every time, which shouldn't happen as [I mentioned here](https://github.com/cockpit-project/cockpit-navigator/issues/298#issuecomment-1976482702).
2. The `files` dependency of the `useMemo` was changing on every render making the memo useless.

### CODE REVIEW

**Comment by jelly** [comment]
Interesting!

So for my one event listener PR I made an event click profile:

After my changes, just measuring a click on a folder:
image

**Comment by jelly** [comment]
Testing with your changes here:

Before:

image

After:

image

**Comment by jelly** [comment]
Splitting Item off to a new function seems like a no-brainer so happy to get that merged :)

File: src/navigator-card-body.jsx
Comment (cockpituous) [inline src/navigator-card-body.jsx:376]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-303-20240306-101459-1e7c613b-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

File: src/navigator-card-body.jsx
Comment (cockpituous) [inline src/navigator-card-body.jsx:376]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-303-20240306-110858-327980e4-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by subhoghoshX** [comment]
@jelly I think it's much more usable now...

I've a few more ideas:
1. Make the Grid <-> Table switch CSS based instead of JS (based on some class, container queries or whatever) that way we can go around the whole re-rendering of components
2. Use better data structures (Map, Set) than the default array or objects. Usually it doesn't matter but in this case there might be some gain.
3. Simply the DOM structure (as garret suggested). I gave it a try but immediately got stuck in semantic html and finding proper colors, I'll leave it to garret :)
4. Come up with a better algo for filter and sort maybe.. I've to check whether the script is slow or rendering..

Before those, maybe we should start merging some of the stuff from here (in separate PRs)? This one is becoming a combination of related, yet unrelated changes..

**Comment by jelly** [comment]
> @jelly I think it's much more usable now...
> 
> I've a few more ideas:
> 
>     1. Make the Grid <-> Table switch CSS based instead of JS (based on some class, container queries or whatever) that way we can go around the whole re-rendering of components
> 
>     2. Use better data structures (Map, Set) than the default array or objects. Usually it doesn't matter but in this case there might be some gain.
> 
>     3. Simply the DOM structure (as garret suggested). I gave it a try but immediately got stuck in semantic html and finding proper colors, I'll leave it to garret :)
> 
>     4. Come up with a better algo for filter and sort maybe.. I've to check whether the script is slow or rendering..
> 
> 
> Before those, maybe we should start merging some of the stuff from here (in separate PRs)? This one is becoming a combination of related, yet unrelated changes..

The separate class for Item is a great one to split out :) I think I noticed the biggest gain there.

**Comment by subhoghoshX** [comment]
> The separate class for Item is a great one to split out :) I think I noticed the biggest gain there.
> 

And with the latest memoization commit it's almost perfect :) feel free to give it a try.

**Comment by jelly** [comment]
Amazing! This makes it really feel a lot faster in when viewing `/lib64`

image

File: src/navigator-card-body.jsx
Comment (jelly) [inline src/navigator-card-body.jsx:362]: The indentation is a bit weird here.

File: src/navigator-card-body.jsx
Comment (jelly) [inline src/navigator-card-body.jsx:295]: Please add to the commit message why this helps with skipping pointless re-drawing.

File: src/app.tsx
Comment (jelly) [inline src/app.tsx:118]: Would be nice to know why this was needed? Or a short comment that this saves selection a new item not re-filter files passed to the FolderView.

File: src/navigator-card-body.jsx
Comment (jelly) [inline src/navigator-card-body.jsx]: How does this help? Selected is still passed to `Item`.

File: src/app.tsx
Comment (jelly) [inline src/app.tsx:70]: This feels a bit overdoing it, splitting a path can not be that slow. Honestly we should moved `selected` somewhere else. But the issue is that we created the sidepanel, so likely we should just throw that into a new Component with the SidePanel.

File: src/navigator-card-body.jsx
Comment (jelly) [inline src/navigator-card-body.jsx:358]: This is stylistic, not related to performance

File: src/navigator-card-body.jsx
Comment (jelly) [inline src/navigator-card-body.jsx:373]: Uhhhh, this is a function component, given the same input it should not re-render, why are we using memo here?

File: src/navigator-card-body.jsx
Comment (jelly) [reply src/navigator-card-body.jsx:358]: Nevermind, this is needed. But the Memo is not.

**Comment by subhoghoshX** [comment]
I see you reviewed this PR.. I was going to create new ones, so the reasons are clear step by step.

File: src/navigator-card-body.jsx
Comment (jelly) [reply src/navigator-card-body.jsx:373]: Except when testing it does not.... that's strange. Also note that selecting an item re-renders an `Item` twice for some reason :thinking:

Comment (subhoghoshX) [reply src/navigator-card-body.jsx:373]: All components re-render whenever the parent component re-renders. I think there's confusing between the terms "re-render" and "updating the dom". "re-render"s always happen for everything. "updading the dom" happens only when there's different out from the component.

**Comment by jelly** [comment]
Also interesting, chromium's devtools can see when the DOM is re-rendered:

Before:

[Screencast from 2024-03-06 14-07-30.webm](https://github.com/cockpit-project/cockpit-navigator/assets/67428/a20a3ceb-9cc6-41aa-bccd-4d979f3d6f1a)

After:
[Screencast from 2024-03-06 14-05-36.webm](https://github.com/cockpit-project/cockpit-navigator/assets/67428/d1288458-78f4-4e08-9e30-05a241be2a36)

So this fixed:
1. DOM re-rendering
2. React comparison being slow? ie. that might explain the Memo.


> I see you reviewed this PR.. I was going to create new ones, so the reasons are clear step by step.

But you can do that here no? I assume the commits don't change?

File: src/navigator-card-body.jsx
Comment (subhoghoshX) [reply src/navigator-card-body.jsx:295]: This method is only used in `Item` component` so it belongs there.

File: src/navigator-card-body.jsx
Comment (subhoghoshX) [reply src/navigator-card-body.jsx:362]: That's how eslint was happy, it doesn't automatically fix it. And I couldn't find any other way.

**Comment by subhoghoshX** [comment]
> > I see you reviewed this PR.. I was going to create new ones, so the reasons are clear step by step.
> 
> But you can do that here no? I assume the commits don't change?

I'll make the commit messages a bit better then, I didn't give much though to it before.

**Comment by jelly** [comment]
> > > I see you reviewed this PR.. I was going to create new ones, so the reasons are clear step by step.
> > 
> > 
> > But you can do that here no? I assume the commits don't change?
> 
> I'll make the commit messages a bit better then, I didn't give much though to it before.

Awesome, I would be very happy to get this merged :) Also as it likely will impact or conflict with some other stuff I have ready/planned like:

https://github.com/cockpit-project/cockpit-navigator/pull/249

File: src/navigator-card-body.jsx
Comment (cockpituous) [inline src/navigator-card-body.jsx:376]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-303-20240306-140502-d545e596-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by subhoghoshX** [comment]
> Awesome, I would be very happy to get this merged :) Also as it likely will impact or conflict with some other stuff I have ready/planned like:
> 
> #249

Thanks, I think it's ready now. Also doesn't React automatically do event delegation?

File: src/navigator-card-body.jsx
Comment (jelly) [reply src/navigator-card-body.jsx:362]: I think you need to manually massage it, but I'd have to take a look. Anyway this is already a much better overall improvement

File: src/app.tsx
Comment (jelly) [reply src/app.tsx:122]: Thanks! I have follow up plans to just move this into NavigatorCard or however it is called, it's only used there so makes no sense to have here.

File: src/navigator-card-body.jsx
Comment (jelly) [reply src/navigator-card-body.jsx]: Ah, I suppose this was the check for if where `selected` == the current directory which was just super confusing.

File: src/navigator-card-body.jsx
Comment (jelly) [inline src/navigator-card-body.jsx:416]: OOo this was a race, nice!

File: src/navigator-card-body.jsx
Comment (jelly) [inline src/navigator-card-body.jsx:373]: I would prefer a comment here in the code, as it was not super obvious. But also fine to merge without.

File: src/navigator-card-body.jsx
Comment (jelly) [reply src/navigator-card-body.jsx:358]: Passing isSelected makes a ton more sense logically and stops a re-render :)

**Review by jelly — APPROVED** [review]

File: src/navigator-card-body.jsx
Comment (subhoghoshX) [reply src/navigator-card-body.jsx:374]: Done!

File: src/navigator-card-body.jsx
Comment (cockpituous) [inline src/navigator-card-body.jsx:377]: These 2 added lines are not executed by any test.  [Details](https://cockpit-logs.us-east-1.linodeobjects.com/pull-303-20240306-165505-26839ef8-fedora-39-devel//Coverage/lcov/github-pr.diff.gcov.html)

**Comment by jelly** [comment]
Thanks for this performance improvement \o/ :rocket:

**Pull request merged**
